### PR TITLE
Document stream validation approach used in testing

### PR DESCRIPTION
**Pull request by vyasr**
## Description

libcudf's stream validation strategy is fairly sophisticated and uses some tooling that is generally unfamiliar to most developers. This PR adds an extended description to the developer docs, along with guidance on how this tooling should be used going forward as move to add streams to more APIs.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/doxygen/developer_guide/TESTING.md [hunk 1]
```diff
@@ -458,3 +458,69 @@ Column comparison functions in the `cudf::test::detail` namespace should **NOT**
...
+libcudf employs a custom-built [preload library
+docs](https://man7.org/linux/man-pages/man8/ld.so.8.html) to validate its internal stream usage (the
+code may be found
+[`here`](https://github.com/rapidsai/cudf/blob/main/cpp/tests/utilities/identify_stream_usage.cpp)).
+This library wraps every asynchronous CUDA runtime API call that accepts a stream with a check to
+ensure that the passed CUDA stream is a valid one, immediately throwing an exception if an invalid
+stream is detected. Running tests with this library loaded immediately triggers errors if any test
+accidentally runs code on an invalid stream.
+
+Stream validity is determined by overloading the definition of cudf's default stream. Normally, in
```
Comment (vuule) [inline cpp/doxygen/developer_guide/TESTING.md:476]: maybe
```suggestion
Stream validity is determined by overloading the definition of libcudf's default stream. Normally, in
```

**Review by vuule — APPROVED** [review]
Nice write-up! 
Left a few non-blocking suggestions.

File: cpp/doxygen/developer_guide/TESTING.md [hunk 2]
```diff
@@ -458,3 +458,69 @@ Column comparison functions in the `cudf::test::detail` namespace should **NOT**
...
+some `detail` function that accepts a stream parameter fails to forward it along and instead
+erroneously calls `cudf::get_default_stream` instead.
+
+In `testing` mode, the library instead overloads `cudf::test::get_default_stream`. This function
+defined in the `cudf::test` namespace enables a more stringent mode of testing. In `testing` mode,
+the preload library instead verifies that all CUDA runtime APIs are instead called using the test
+namespace's default stream. This distinction is important because cudf internals never use
+`cudf::test::get_default_stream`, so this stream value can only appear internally if it was provided
+ot a public API and forwarded properly all the way down the call stack. While `testing` mode is more
+strict `cudf` mode, it is also more intrusive. `cudf` mode can operate with no changes to the
```
Comment (davidwendt) [inline cpp/doxygen/developer_guide/TESTING.md:500]:
```suggestion
to a public API and forwarded properly all the way down the call stack. While `testing` mode is more
strict than `cudf` mode, it is also more intrusive. `cudf` mode can operate with no changes to the
```

**Review by davidwendt — APPROVED** [review]

**Pull request merged**